# 4. Data Wrangling and Subsetting
## Table of contents
***

**`1. Import libraries`**<br/>

**`2. Create path and import data`**<br/>

**`3. Wrangling data`** <br/>
- _3.1 Change data types_<br/>
- _3.2 Rename columns_<br/>
- _3.3 Drop columns_<br/>

 
**`4. Wrangling department dataframe`** <br/>
- _4.1 Transposing department dataframe_ <br/>
- _4.2 Change header of the department dataframe_ <br/>
- _4.3 Create data dictionary_ <br/>

**`5. Locate the busiest hour of day`**<br/>

**`6. Understand breakfast item sales`**<br/>
- _6.1 Create a subset for breakfast item_<br/>
- _6.2 Create a subset for dinner party item_<br/>
- _6.3 Understand the specific user behavior_<br/>

**`7. Export data`**
***

### 1. Import Libraries

In [54]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 2. Create and import data

In [55]:
# Create path
path = r'/Users/bonnieandbanana/Desktop/A New Career/CareerFoundry/Data Immersion/A4_Python/ Instacart Basket Analysis Project (IC)'

# Orders data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

# Products data
df_prods = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'products.csv'), index_col = False)

# Departments data
df_dep = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'departments.csv'), index_col = False)

### 3. Wrangling data
#### _3.1 Change data types_

In [56]:
# Changing data type for 'user_id' variable from numerical to string
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [57]:
# Changing data type for 'order_id variable from numerical to string'
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [58]:
# Check results
df_ords['order_id'].dtype
df_ords['user_id'].dtype

dtype('O')

**Observation**

object type in pandas = string in Python

#### _3.2 Rename columns_

In [59]:
df_ords.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = True)
df_ords.rename(columns= {'order_dow' : 'order_day_of_the_week'}, inplace = True)

In [60]:
df_ords.head()

order_id user_id eval_set  order_number  order_day_of_the_week  \
0  2539329       1    prior             1                      2   
1  2398795       1    prior             2                      3   
2   473747       1    prior             3                      3   
3  2254736       1    prior             4                      4   
4   431534       1    prior             5                      4   

   order_hour_of_day  days_since_last_order  
0                  8                    NaN  
1                  7                   15.0  
2                 12                   21.0  
3                  7                   29.0  
4                 15                   28.0

#### _3.3 Drop columns_

In [61]:
df_ords = df_ords.drop(columns = ['eval_set'])

In [78]:
# Check results
df_ords.head()

order_id user_id  order_number  order_day_of_the_week  order_hour_of_day  \
0  2539329       1             1                      2                  8   
1  2398795       1             2                      3                  7   
2   473747       1             3                      3                 12   
3  2254736       1             4                      4                  7   
4   431534       1             5                      4                 15   

   days_since_last_order  
0                    NaN  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0

**Observation**

After careful consideration, the unnecessary column of 'eval_est' has been dropped succesfully from the dataframe.

## 4. Wrangling department dataframe 

#### _4.1 Transposing the department df_

In [63]:
# Transposing the department data
df_dep_t = df_dep.T

#### _4.2 Change header of the department df_

In [80]:
# Step1: adding an index
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [65]:
# Step 2: create a new header
new_header = df_dep_t.iloc[0]

In [66]:
# Step 3: create new dataframe to discard the old header
df_dep_t_new = df_dep_t[1:]

In [67]:
# Step 4: set the header rows as the df header
df_dep_t_new.columns = new_header

In [81]:
# Check the final result of header changing
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

#### _4.3 Create data dictionary_

In [82]:
# Set the new df as data dictionary
data_dict = df_dep_t_new.to_dict('index')

In [83]:
# Check result of the data dictionary
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

**Observation**

The data dictionary is created using the index number to represent a single entry which helps Python locate the information in the dictionary efficiently.

### 5. Locate the busiest hour of day

In [86]:
# Find the busiest hour for placing orders: 10 o'clock
df_ords['order_hour_of_day'].value_counts (dropna= False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

**Observation**

The busiest hour of order being placed is 10 o'clock and the least busy hour is 3 o'clock in the morning. It starts to get busy from 8am and peak at 10am. From 10am till 17pm, the orders slightly decrease but still remain considerably high, however, orders drastically decrease after 18:00pm.

In [85]:
# Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe
data_dict.get('4')

{'department': 'produce'}

### 6. Understand breakfast item sales

#### _6.1 Create a subset for breakfast item_

In [97]:
# Create a subset for breakfast item, identify the department id first
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

**Observation**

The index represents breakfast department is '14'.

In [106]:
# Create the subset for breakfast item
df_prods_breakfast = df_prods.loc[df_prods['department_id'].isin([14])]

In [107]:
# Check result
df_prods_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

#### _6.2 Create a subset for dinner party item_

In [109]:
# Create 'dinner party' subset
df_prods_dinner = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [110]:
# Check result
df_prods_dinner.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

**Observation**

The dataframe created above for dinner party could provide the sales team a good glance to start their analysis.

In [112]:
# Find out how many rows are in the dataframe last created
df_prods_dinner.shape

(7650, 5)

**Observation**

The shape of dataframe 'df_dinner_party' shows 7650 rows and 5 columns, meaning 7650 products are included in the list.

#### _6.3 Understand the specific user behavior_

In [113]:
# Extracting all information about user_1
df_user_1 = df_ords.loc[df_ords['user_id']=='1']
df_user_1

order_id user_id  order_number  order_day_of_the_week  order_hour_of_day  \
0   2539329       1             1                      2                  8   
1   2398795       1             2                      3                  7   
2    473747       1             3                      3                 12   
3   2254736       1             4                      4                  7   
4    431534       1             5                      4                 15   
5   3367565       1             6                      2                  7   
6    550135       1             7                      1                  9   
7   3108588       1             8                      1                 14   
8   2295261       1             9                      1                 16   
9   2550362       1            10                      4                  8   
10  1187899       1            11                      4                  8   

    days_since_last_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
5                    19.0  
6                    20.0  
7                    14.0  
8                     0.0  
9                    30.0  
10                   14.0

In [115]:
# Provide basic stats of the behavior of user_1
df_user_1.describe()

order_number  order_day_of_the_week  order_hour_of_day  \
count     11.000000              11.000000          11.000000   
mean       6.000000               2.636364          10.090909   
std        3.316625               1.286291           3.477198   
min        1.000000               1.000000           7.000000   
25%        3.500000               1.500000           7.500000   
50%        6.000000               3.000000           8.000000   
75%        8.500000               4.000000          13.000000   
max       11.000000               4.000000          16.000000   

       days_since_last_order  
count              10.000000  
mean               19.000000  
std                 9.030811  
min                 0.000000  
25%                14.250000  
50%                19.500000  
75%                26.250000  
max                30.000000

#### Observation:
This user only places orders from Sunday to Wednesday, never from Thrusday to Saturday. 
More than half of the time, he places orders at or before 8am and the orders are placed in average every 19 days. 
He never spend more than 30 days without placing orders, and once he placed 2 orders on the same day.

### 7. Export data

In [117]:
# Export orders dataframe
df_ords.to_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'))

In [119]:
# Export department dataframe
df_dep_t_new.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'))